lib/std/smartptrs.nim

```diff
@@ -0,0 +1,211 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+
+
+when defined(nimExperimentalSmartptrs) or defined(nimdoc):
+  import std/isolation
+
+
+  type
+    UniquePtr*[T] = object
+      ## Non copyable pointer to a value of type `T` with exclusive ownership.
+      val: ptr T
+
+  proc `=destroy`*[T](p: var UniquePtr[T]) =
+    if p.val != nil:
+      `=destroy`(p.val[])
+      when compileOption("threads"):
+        deallocShared(p.val)
+      else:
+        dealloc(p.val)
+
+  proc `=copy`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+    ## The copy operation is disallowed for `UniquePtr`, it
+    ## can only be moved.
+
+  proc newUniquePtr[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+    ## Returns a unique pointer which has exclusive ownership of the value.
+    when compileOption("threads"):
+      result.val = cast[ptr T](allocShared(sizeof(T)))
+    else:
+      result.val = cast[ptr T](alloc(sizeof(T)))
+    # thanks to '.nodestroy' we don't have to use allocShared0 here.
+    # This is compiled into a copyMem operation, no need for a sink
+    # here either.
+    result.val[] = val.extract
+    # no destructor call for 'val: sink T' here either.
+
+  template newUniquePtr*[T](val: T): UniquePtr[T] =
```

Comment:
as mention in https://github.com/nim-lang/Nim/pull/18450#discussion_r666008077 for the other PR, this is what's needed:
```nim
proc newUniquePtr*[T](U: typedesc[T]): UniquePtr[T] =
  result.val = cast[ptr T](allocShared(sizeof(T)))
```
which also ends up producing correct results for C++

read the whole thread starting here: https://github.com/nim-lang/Nim/pull/18450#discussion_r665751874

Reply:
I don't understand. Don't you want the unique pointer to point to some valid object? Your proc doesn't do the same.

Reply:
I'm allocating on the heap only once instead of allocating in caller scope (eg on stack) and then allocating on heap and copying; it's more efficient and analogous to C++ placement new on which C++ `make_unique` relies, and is also more correct wrt destructors being called as you can see in https://github.com/nim-lang/Nim/pull/18450#discussion_r666008077

Caller can then, if needed, assign values in the heap allocated object, eg:
```nim
import std/smartptrs

type ManagedFile = object
  name: string
  cfile: File

proc `=destroy`(a: var ManagedFile) =
  echo ("in =destroy", a.name)
  close(a.cfile)
  a.cfile = nil

proc main()=
  block:
    let p = newUniquePtr(ManagedFile)
    p[].name = currentSourcePath
    p[].cfile = open(p[].name, fmRead)
    echo p[].cfile.readAll.len
  echo "after"
main()
```

prints:
```
381
("in =destroy", "/Users/timothee/git_clone/nim/timn/tests/nim/all/t12621.nim")
after
```

Reply:
@timotheecour its already taken care of https://github.com/nim-lang/threading/blob/master/threading/smartptrs.nim#L48 Its not more efficient though, you should use the sink one as I said in https://github.com/nim-lang/Nim/pull/17333#issuecomment-870486146

Reply:
```nim

    p[].name = currentSourcePath
    p[].cfile = open(p[].name, fmRead)

```

is not as efficient as direct inplace construction. Not because of the lack of moves, but because the compiler doesn't know that in `p[]` there is nothing to destroy yet.

Reply:
```nim
    p[].name = currentSourcePath
    p[].cfile = open(p[].name, fmRead)
```

...also it is a loophole, as it bypasses isolate.

Reply:
> is not as efficient as direct inplace construction. Not because of the lack of moves, but because the compiler doesn't know that in p[] there is nothing to destroy yet.

@Araq please provide an example to demonstrate this.
Here's one, showing that the way I proposed is 3.4X faster, where the speedup depends on the size of the object; you'll get smaller speedups for smaller objects but always at least as fast.

nim r -d:danger --gc:arc -d:case2a -d:nimExperimentalSmartptrs t12621.nim
  0.058782
nim r -d:danger --gc:arc -d:case2b -d:nimExperimentalSmartptrs t12621.nim
  0.20173

```nim
when true:
  import std/smartptrs
  import std/times
  type Foo = object
    a1: int
    a: array[1000, int]

  proc fun(i: int): int =
    when defined(case2a):
      let p = newUniquePtr(Foo)
      p[].a1 = i
    when defined(case2b):
      let p = newUniquePtr(Foo(a1: i))
    result += cast[int](p)

  proc main =
    let t = cpuTime()
    let n = 1000000
    var tot = 0
    for i in 0..<n:
      tot += fun(i)
    let t2 = cpuTime()
    echo (t2 - t, tot)
  main()
```

Furthermore, the way I propose allows client code to use newUniquePtr with C++ objects with placement new constructor (and no moves, you construct the object in place directly), whereas it's not possible with existing API.

And correctness wrt destructors is the other obvious advantage of the proposed approach.

> ...also it is a loophole, as it bypasses isolate.

@planetis-m I don't follow, the loophole is already there, regardless of whether `proc newUniquePtr*[T](U: typedesc[T]): UniquePtr[T]` is added:

```nim
when defined case3:
  import std/smartptrs
  type Bar = ref object
  type Foo = object
    bar: Bar
  proc main()=
    let bar = Bar()
    block:
      let p = newUniquePtr(Foo())
      p[].bar = bar
    echo "after"
  main()
```

Reply:
> Here's one, showing that the way I proposed is 3.4X faster, where the speedup depends on the size of the object; you'll get smaller speedups for smaller objects but always at least as fast.

But that's measuring artifacts of the current implementation, in particular NVRO (which is unfortunately currently fragile). In theory it's a pessimization as you mitigated the compiler's ability to reason about the code.

Reply:
I still don't understand the reasoning here; in current code, you:
1 construct an object Foo in the stack
2 allocate sizeof(Foo) on the heap
3 copy Foo

with proposed code you:
1 allocate sizeof(Foo) on the heap
2 construct in-place

in all possible scenarios, it should be at least as fast, and usually faster; eg it allows placement-new with C++, or simply just constructing in-place the fields you care about updating from their default value (if some of the memory is to remain 0-initialized, you don't pay for it); you also don't have extra ctor/dtor to worry about, in particular for C++ objects.

## n element version
Finally, this generalizes in a straightforward way to C++'s array version of make_unique, which we should add at some point; it allows dynamic allocation of n objects without having to use a seq (eg so you can interface with C++ or avoid relying on gc for allocating n elements); there are 2 approaches to consider:
* same as C++, make the n implicit (user code need to pass that value around somewhere else)
```nim
proc newUniquePtr*[T](U: typedesc[T], n = 1): UniquePtr[T] =
  result.val = cast[ptr T](allocShared(sizeof(T) * n))
```

* add a `UniquePtrN` type with an extra field `n: int`
```nim
type
  UniquePtrN*[T] = object
    val: ptr T
    n: int
proc newUniquePtrN*[T](U: typedesc[T], n = 1): UniquePtrN[T] =
  result.val = cast[ptr T](allocShared(sizeof(T) * n))
  result.n = n
```

Reply:
> I don't follow, the loophole is already there,

I am just pointing it out. I had totally missed it before.

> n element version

Why not use a seq instead?

Reply:
> I still don't understand the reasoning here; in current code, you: ...

Your're correct, however, the code is still not as efficient as it could be if we had proper "in-place" construction. I think we do proper in-place construction for the construct `toSinkParam ObjConstr(x: x, y: y, ...)`.

Reply:
for c++ objects, it should already be possible today without a language extension (via importcpp of `new(buf) T`), although i haven't tried yet
for nim objects, it shouldn't be hard to extend `semObjConstr` to take an optional input buffer address, eg:
```nim
var p: ptr Foo = ...
Foo.new(p, a1: 1, a2: 2) # explicit syntax
p[] = Foo(a1: 1, a2: 2) # syntax sugar possible?
```

in any case, these can be added later, and `proc newUniquePtrN*[T](U: typedesc[T]): UniquePtrN[T] =` can be introduced before those exist; code would just get auto-optimized once `p[] = Foo(a1: 1, a2: 2)` starts working as in-place construction